Add contributing guide (#804)
stellar/freighter-mobile: CONTRIBUTING.md

# Contributing to Freighter Mobile

Non-custodial Stellar wallet for iOS and Android. Built with React Native,
Zustand, and React Navigation.

For the Stellar organization's general contribution guidelines, see the
[Stellar Contribution Guide](https://github.com/stellar/.github/blob/master/CONTRIBUTING.md).

## Prerequisites

| Tool           | Version       | Install                                                           |
| -------------- | ------------- | ----------------------------------------------------------------- |

> | Node.js        | >= 20         | [nodejs.org](https://nodejs.org) or `nvm install 20`              |

Reviewer: I'm on Node `v22.14.0` so I'd suggest we use `nvm install 22` as command sample and also double check if the min required version is in fact v20 instead of v22

| Yarn           | 4.10.0        | `corepack enable && corepack prepare yarn@4.10.0 --activate`      |

> | Ruby           | >= 2.6.10     | [rbenv](https://github.com/rbenv/rbenv) or [rvm](https://rvm.io/) |

Reviewer: just noting that I'm on `ruby 3.1.4p223`

bundle install        # Ruby deps (Fastlane, CocoaPods)
yarn install          # Node deps + auto-runs postinstall (Husky, polyfills, pods)
cp .env.example .env  # Then fill in values (see below)
```

### Environment Variables

Copy `.env.example` to `.env` and fill in the values. The `.env` file must
**never** be committed.

**Required — app won't build or function without these:**

| Variable                            | How to set up                                                                                                                                                                                                                        |
| ----------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `FREIGHTER_BACKEND_V1_DEV_URL`      | `https://freighter-backend-dev.stellar.org/api/v1`. If unavailable, use `-stg` or `-prd`. Or run your own from [stellar/freighter-backend](https://github.com/stellar/freighter-backend)                                             |

> | `FREIGHTER_BACKEND_V2_DEV_URL`      | `https://freighter-backend-v2-dev.stellar.org/api/v1`. If unavailable, use `-stg` or prod (`freighter-backend-v2.stellar.org`). Or run your own from [stellar/freighter-backend-v2](https://github.com/stellar/freighter-backend-v2) |

Reviewer: ```suggestion
| `FREIGHTER_BACKEND_V1_DEV_URL`      | Run your own backend from [stellar/freighter-backend](https://github.com/stellar/freighter-backend)                                             |
| `FREIGHTER_BACKEND_V2_DEV_URL`      | Run your own backend-v2 from [stellar/freighter-backend-v2](https://github.com/stellar/freighter-backend-v2) |
```

> | `WALLET_KIT_PROJECT_ID_DEV`         | Create a free project at [dashboard.reown.com](https://dashboard.reown.com) — sign up, create a new project (type: Wallet), copy the Project ID                                                                                      |

Reviewer: ```suggestion
| `WALLET_KIT_PROJECT_ID_DEV`         | Create a free project at [dashboard.walletconnect.com](https://dashboard.walletconnect.com) — sign up, create a new project (type: Wallet), copy the Project ID                                                                                      |
```

> | `WALLET_KIT_MT_NAME_DEV`            | Your project name from the Reown dashboard                                                                                                                                                                                           |

Reviewer: ```suggestion
| `WALLET_KIT_MT_NAME_DEV`            | Your project name from the WalletConnect dashboard                                                                                                                                                                                           |
```

| `WALLET_KIT_MT_REDIRECT_NATIVE_DEV` | Deep link scheme matching your dev bundle ID                                                                                                                                                                                         |
| `ANDROID_DEBUG_KEYSTORE_PASSWORD`   | Android Studio's default: `android`                                                                                                                                                                                                  |
| `ANDROID_DEBUG_KEYSTORE_ALIAS`      | Android Studio's default: `androiddebugkey`                                                                                                                                                                                          |
| `ANDROID_DEV_KEYSTORE_PASSWORD`     | Generate your own keystore: `mkdir -p android/keystores && keytool -genkey -v -keystore android/keystores/dev-release.keystore -alias dev -keyalg RSA -keysize 2048 -validity 10000`, then use the password you set                  |

> | `ANDROID_DEV_KEYSTORE_ALIAS`        | The alias you chose when generating the keystore (e.g., `dev`)                                                                                                                                                                       |

Reviewer: I don't think those 2  `ANDROID_DEV_` keys are required to run locally, we only need those in release mode

**Optional — features degrade gracefully without these:**

| Variable                                     | Notes                                                                                      |
| -------------------------------------------- | ------------------------------------------------------------------------------------------ |
| `AMPLITUDE_API_KEY`                          | Auto-disabled in `__DEV__` mode — not needed for local dev                                 |
| `SENTRY_DSN`                                 | Leave empty — errors log to console instead                                                |

> | `FREIGHTER_BACKEND_*_STG_URL` / `*_PROD_URL` | Only needed for staging/prod builds. Same public endpoints with `-stg` or `-prd` subdomain |

Reviewer: ```suggestion
| `FREIGHTER_BACKEND_*_STG_URL` / `*_PROD_URL` | Only needed for staging/prod builds. Run your own backend from [stellar/freighter-backend](https://github.com/stellar/freighter-backend) and backend-v2 from [stellar/freighter-backend-v2](https://github.com/stellar/freighter-backend-v2)  |
```

See `.env.example` for the full list with inline comments.

### Run the App

```bash
yarn ios              # iOS simulator (dev)
yarn android          # Android emulator (dev)
yarn start            # Metro bundler only (starts automatically with above)
```

Dev and prod builds use separate bundle IDs (`org.stellar.freighterdev` /
`org.stellar.freighterwallet`) and can coexist on the same device. See
`package.json` for all run commands including device and release variants.

## Key Commands

```bash
yarn test             # Jest unit tests
yarn check            # All checks (TypeScript + ESLint + Prettier)
yarn fix              # Auto-fix lint + format
yarn lint:translations  # Check for missing i18n keys
yarn test:e2e:ios <flow>     # Maestro e2e (iOS)
yarn test:e2e:android <flow> # Maestro e2e (Android)
```

**Cleaning builds** (escalation order):

```bash
yarn start-c          # Clear Metro cache
yarn pod-install      # Reinstall CocoaPods

> yarn node-c-install   # Remove node_modules + reinstall

Reviewer: I'd also add "yarn gradle-clean" here for convenience

# Contributing to Freighter Mobile

Non-custodial Stellar wallet for iOS and Android. Built with React Native,
Zustand, and React Navigation.

For the Stellar organization's general contribution guidelines, see the
[Stellar Contribution Guide](https://github.com/stellar/.github/blob/master/CONTRIBUTING.md).

## Prerequisites

> | Tool           | Version       | Install                                                                                     |

Reviewer: We should probably add Cocoapods to the list too. I'm on Cocoapods 1.15.2.
We also need to update the quick-start-guide with it

Reviewer: The `Podfile.lock` file uses `COCOAPODS: 1.15.2` so we should probably steer people to use the same

| `WALLET_KIT_PROJECT_ID_DEV`         | Create a free project at [dashboard.walletconnect.com](https://dashboard.walletconnect.com) — sign up, create a new project (type: Wallet), copy the Project ID |
| `WALLET_KIT_MT_NAME_DEV`            | Your project name from the WalletConnect dashboard                                                                                                              |
| `WALLET_KIT_MT_DESCRIPTION_DEV`     | Your project description                                                                                                                                        |
| `WALLET_KIT_MT_URL_DEV`             | Your project URL                                                                                                                                                |
| `WALLET_KIT_MT_ICON_DEV`            | Your project icon URL                                                                                                                                           |

> | `WALLET_KIT_MT_REDIRECT_NATIVE_DEV` | Deep link scheme matching your dev bundle ID                                                                                                                    |

Reviewer: ```suggestion
| `WALLET_KIT_MT_REDIRECT_NATIVE_DEV` | Deep link matching what you've configured on WalletConnect dashboard                                                                                                                    |
```

Husky runs on every commit:

1. `lint-staged` — ESLint fix + Prettier on staged files
2. `yarn test` — full unit test suite
3. `yarn lint:ts` — TypeScript type check

All must pass before the commit succeeds.

## Testing

**Unit tests:** Jest with `@testing-library/react-native`. Tests in
`__tests__/`, mocks in `__mocks__/`.

**E2E tests:** Maestro flows in `e2e/flows/` — see the [e2e docs](e2e/docs/) for
setup, writing tests, and debugging.

**Before submitting a PR:** `yarn test` + `yarn lint:ts` must pass. Test on both
iOS and Android, including small screens.

## Adding Features

### Adding a new screen

1. Create the screen component in `src/components/screens/`
2. Add the route to `src/config/routes.ts`
3. Register in the appropriate navigator in `src/navigators/`
4. Add translations for all user-facing strings in `src/i18n/`
5. Add tests in `__tests__/`

### Adding a new Zustand store

1. Create the store file in `src/ducks/` (e.g., `src/ducks/myFeature.ts`)
2. Define the store using Zustand's `create` with typed state and actions
3. Export a hook for consuming the store (e.g., `useMyFeatureStore`)
4. Add tests in `__tests__/ducks/`

### Platform-specific code

> - Use `Platform.OS` checks for small differences

Reviewer: I think we should default to using our existing `isIOS` and `isAndroid` helpers here

- Use `.ios.tsx` / `.android.tsx` file extensions for larger divergences
- Test on **both platforms** — especially native module interactions, gestures,
  and keyboard behavior

## Running the Backend Locally

The app connects to two separate backend services:

- **V1**
  ([stellar/freighter-backend](https://github.com/stellar/freighter-backend),

>   TypeScript) — balances, subscriptions, feature flags, notifications

Reviewer: We should remove `feature flags` from here as the /feature-flags endpoint is only being used on extension. I think `subscriptions` and `notifications` are also probably coming from extension. Could you please review V1 usage on mobile and update the description here?

- **V2**
  ([stellar/freighter-backend-v2](https://github.com/stellar/freighter-backend-v2),

>   Go) — collectibles, RPC health, ledger keys, protocols

Reviewer: this one also needs to be reviewed and updated

## Pull Requests

- Branch from `main` using your initials + description: `lf-feature-name`,
  `cg-fix-token-display`
- Commit messages: start with action verb (`Add`, `Fix`, `Update`, `Improve`)
- No mixed concerns — keep refactoring separate from features
- Include before/after screenshots for UI changes
- Follow the full checklist in the
  [PR template](.github/pull_request_template.md)

**CI runs on every PR:** unit tests (`test.yml`), iOS e2e (`ios-e2e.yml`),
Android e2e (`android-e2e.yml`). All must pass.

## Security

Freighter handles private keys and signs transactions. When contributing:

- **Never log or expose** private keys, seed phrases, or passwords
- **Use `react-native-keychain`** (iOS Keychain / Android Keystore) for secrets
  — never AsyncStorage
- **Validate all external data** — WalletConnect payloads, API responses, deep
  links
- **Don't weaken Blockaid** transaction scanning or `jail-monkey` jailbreak
  detection
- **Report vulnerabilities** via the
  [Stellar Security Policy](https://github.com/stellar/.github/blob/master/SECURITY.md)
  — not public issues

See [`docs/auth_flow_diagram.md`](docs/auth_flow_diagram.md) for the full auth

> security model.

Reviewer: it has been 11 months since the last time this diagram has been updated, I'd either avoid mentioning it on this contributing doc or thoroughly review and update the auth diagram doc to avoid confusion

## Further Reading

| Topic                      | Location                                                                                      |
| -------------------------- | --------------------------------------------------------------------------------------------- |

> | Auth architecture          | [`docs/auth_flow_diagram.md`](docs/auth_flow_diagram.md)                                      |

Reviewer: same here
